### ISSUE TITLE
Dagger fails if invalid or expired GHCR token is present in ~/.docker/config.json file

### ISSUE THREAD

**Issue by marcosnils** [dagger/dagger#4381]
## Problem

Since dagger pulls the engine image from `ghcr`, it'll try to use a token if present in the ~/.docker/config,json file and if invalid or expired, it'll fail with an error similar to this one: 

```
 go run main.go 
panic: EOF: Error: resolving image digest: GET https://ghcr.io/token?scope=repository%3Adagger%2Fengine%3Apull&service=ghcr.io: DENIED: denied
```

Which can be difficult for users to troubleshoot.  

## Alternatives

1 Do nothing.
2 Better error handling and tell the user to fix the token issues themselves.
3 If token issue, try "without" a token as long as the default engine image is used (no _EXPERIMENTAL_DAGGER_RUNNER_HOST).

Thoughts?

I think we could start with 2 since 3 will require some more custom token code handling as all the pulling logic is mostly being handled by `docker run` today AFAIK. Please correct me if I'm wrong @sipsma 

cc @shykes @sipsma

**Comment by samalba** [comment]
Until we had proper support for authenticated pull: #4242 (instead of relying on docker auth), we should at least display a proper error message. I would discuss/implement the rest as part of #4242.

Do we have a record of users impacted? Raise the priority nonetheless.

**Comment by marcosnils** [comment]
Thx for the reply Sam. As you know, the engine image pull and the `from` step follow two very different code paths to perform that action. Just bringing this up since #4242 does not refer to this first scenario which is the one that triggered this issue. 

Shall we expand the scope of #4242 to also cover this case? In the meantime I'll just add some basic error handling for this particular case

> Do we have a record of users impacted? Raising the priority nonetheless.

I was an impacted user :). I was trying to use dagger for something today and took me a while to find the reason why it wasn't working.

**Comment by samalba** [comment]
I agree it can be solved as part of #4242, my point was that until auth is implemented natively in dagger, this should at least be handled from an error handling perspective when the token expires, to avoid confusion.

**Comment by dolanor** [comment]
Impacted as well today.

**Comment by samalba** [comment]
@dolanor @TomChv If one of you have some bandwidth, could you take a look? In particular not sure if it should be part of #4242 or if there is a stopgap for handling this error.

**Comment by grouville** [comment]
#### Context

It took me some time to figure out how to reproduce this error.

My `~/docker/config.json` file was relying on computer credentials:
```json
{
	"auths": {
		"docker.com": {},
		"ghcr.io": {
                       "auth": "Z3JvdXZpbGxlOndyb25nCg==",
                 },
		"https://index.docker.io/v1/": {},
	},
	"credsStore": "desktop",
	"experimental": "disabled",
	"stackOrchestrator": "swarm"
}
```

The above works because docker relies on my credsStore.

To reproduce: 
```json
{
        "auths": {
                "ghcr.io": {
                        "auth": "Z3JvdXZpbGxlOndyb25nCg=="
                }
        }
}
```

#### PR on the way, impacting

> @dolanor @TomChv If one of you have some bandwidth, could you take a look? In particular not sure if it should be part of #4242 or if there is a stopgap for handling this error.

This will be impacted by https://github.com/dagger/dagger/pull/4489 (issue: https://github.com/dagger/dagger/issues/4484). 

New behavior after merge: whenever the pull fails, it will first try to see if a local engine is currently running and use it if it present. Otherwise it will fail.

##### Solution

After checking, it shall not be part of #4242, as it touches the `imageprovider` part of our codebase: https://github.com/dagger/dagger/blob/main/internal/engine/docker.go#L48-L52

As it was a pretty niche use-case (on my side). The most logical solution is to parse the error message and make it clearer ? (maybe by checking for `DENIED` in error message)

**Comment by marcosnils** [comment]
> After checking, it shall not be part of #4242, as it touches the `imageprovider` part of our codebase: https://github.com/dagger/dagger/blob/main/internal/engine/docker.go#L48-L52

This was my initial thought here: https://github.com/dagger/dagger/issues/4381#issuecomment-1382606068

**Comment by grouville** [comment]
@laupse, would you be interested in contributing to this one too ? As your PR https://github.com/dagger/dagger/pull/4489 seems to be in a mergeable state and fixing this one is just managing the `DENIED` use-case to orient the user ?

**Comment by laupse** [comment]
> @laupse, would you be interested in contributing to this one too ? As your PR #4489 seems to be in a mergeable state and fixing this one is just managing the [`DENIED` use-case](https://github.com/dagger/dagger/pull/4489#issuecomment-1410636691) to orient the user ?
> 
> > (Don't want to force you at all ahah, it just seems to be a logical follow-up)

I can ofc. Correct me if i wrong, we want to give a nice message indicating to check the docker creds if we encounter a `DENIED` ?

**Comment by grouville** [comment]
> I can ofc. Correct me if i wrong, we want to give a nice message indicating to check the docker creds if we encounter a `DENIED` ?

Yes exactly, that seems to be an acceptable solution given the scope of this issue 😇

**Comment by grouville** [comment]
Closed by #4489
